scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -39,119 +41,87 @@ class ScaleOutRelayService(
-          _ <- woRuleRepository.delete(ruleSetup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleSetup.id.value} failed")
-        } yield ()
-      }
-      _ <- woRuleRepository.create(ruleDistribPolicy, modId, actor, reason).catchAll { err =>
-        val e = Chained(s"Trying to remove residual object : rule ${ruleDistribPolicy.id.value}", err)
-        PluginLogger.error(e)
-        for {
-          _ <- woLDAPNodeRepository.deleteNode(nodeInf.node, modId, actor, reason).chainError(s"Rollback promote to relay failed : removed ${nodeInf.node} failed")
-          _ <- woLDAPNodeGroupRepository.deletePolicyServerTarget(ruleTarget).chainError(s"Rollback promote to relay failed : removed rule target ${ruleTarget.target} failed")
-          _ <- woLDAPNodeGroupRepository.delete(nodeGroup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed node group ${nodeGroup.id.value} failed")
-          _ <- woDirectiveRepository.deleteActiveTechnique(activeTechniqueId, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed active technique ${activeTechniqueId.value} failed")
-          _ <- woDirectiveRepository.deleteActiveTechnique(activeTechniqueIdCommon, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed active technique ${activeTechniqueIdCommon.value} failed")
-          _ <- woRuleRepository.delete(ruleSetup.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleSetup.id.value} failed")
-          _ <- woRuleRepository.delete(ruleDistribPolicy.id, modId, actor, Some("Promote node to relay rollback")).chainError(s"Rollback promote to relay failed : removed rule  ${ruleDistribPolicy.id.value} failed")
-        } yield ()
-      }
+      _ <- woLDAPNodeRepository.createNode(nodeInf.node, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.createPolicyServerTarget(ruleTarget, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.create(nodeGroup, categoryId, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueId, directDistribPolicy, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueIdCommon, commonDirective, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleSetup, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleDistribPolicy, modId, actor, reason)
     } yield {
       nodeInf
     }
+
+    promote.catchAll { err =>
+      PluginLogger.error(err)
+      IO.effect(rollbackPromote(ruleTarget, ruleSetup, ruleDistribPolicy, nodeInf, nodeGroup, activeTechniqueId, activeTechniqueIdCommon, modId, actor))
+    }
```

Comment:
I don't know if it is the right way to do this : If an error occure when we create an object, perform a rollback

Reply:
It's the right way

But you Should use 

ZioRuntime.runNow(rollbackPromote)

Reply:
in fact err I think you still want to fail and return an error and not a success

Reply:
you just want to map on error with asError (i think the named changed on master branch

Reply:
No, he is in a ZIO, you should not do `runNow`. 
Also, check that the `PluginLogger.error` is pure (a ZIO). If not, used a pure version of the logger. And then, compose it with `PluginLogger.error(err) *> rollbackPromote`.
Rollback promote must be a `ZIO[something, ]`, you should never have `Something[ZIO]`. ie: ZIO must be the outer most structure, else it means you missed a composition somewhere with one of `ZIO` combinators on composite structure like `List`

Reply:
The new name is mapError

Reply:
But it's not a catchAll here it's a mapError, we still want the result to be an error not change our result or else if we have an error, and the rollback is ok you will have a success result and not an error anymore